scroll-tech/rollup-explorer-backend
# feat: add scheduled jobs `last_finalized_monitor` and `precommitted_monitor`

Close https://github.com/scroll-tech/devops/issues/9

### Summary

1. Add a scheduled job `precommitted_monitor` which repeated in every 2 minutes. It runs as:
- Get current `precommitted` IDs from DB.
- Get previous (old) `precommitted` IDs from cache (saving before 2 minutes at previous running).
- Check if any old IDs in new IDs (intersection), notify these IDs to the Slack channel.
- Save new IDs to cache.


2. Add a scheduled job `last_finalized_monitor` which repeated in every 30 minutes. It runs as:
- Get last `finalized` ID from DB.
- Get previous (old) `next-committed` ID from cache (saving before 30 minutes at previous running).
- If `old_committed_id == last_finalized_id + 1` notify this `committed` ID to the Slack channel. Since it has already been saved in cache for 30 minutes (at previous running).
- Get new `next-committed` ID from DB and save to cache.


### Test

<img width="600" alt="1" src="https://user-images.githubusercontent.com/31645658/188352353-df852468-18ae-4ad7-be0d-d33be380f02b.png">

<img width="600" alt="2" src="https://user-images.githubusercontent.com/31645658/188352369-81af6e6f-37b9-4474-97e3-76edd44db447.png">

Both `precommitted_monitor` and `last_finalized_monitor` could be tested manually as:

1. Request a Slack incoming webhook for test (may select your personal channel, for me it is `Steven Gu`) in https://scrollco.slack.com/apps/new/A0F7XDUAZ-incoming-webhooks.

2. Update `MONITOR_LAST_FINALIZED_INTERVAL_SECS` or `MONITOR_PRECOMMITTED_INTERVAL_SECS` to `10` seconds in file `.env`.

3. Import the corresponding test SQL - `db/tests/test_last_finalized_monitor.sql` or `test_precommitted_monitor.sql`.

4. Start the application (`cargo run` or `make start`). You will see the Slack notification soon.
